x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/audit/logfile/LoggingAuditTrail.java

```diff
@@ -416,53 +459,70 @@ public void tamperedRequest(String action, TransportMessage message) {
                 if (indices.isPresent()) {
-                    logger.info("{}[transport] [tampered_request]\t{}, action=[{}], indices=[{}], request=[{}]{}",
-                            localNodeInfo.prefix, originAttributes(threadContext, message, localNodeInfo), action,
-                            arrayToCommaDelimitedString(indices.get()), message.getClass().getSimpleName(), opaqueId());
-                } else {
-                    logger.info("{}[transport] [tampered_request]\t{}, action=[{}], request=[{}]{}", localNodeInfo.prefix,
-                            originAttributes(threadContext, message, localNodeInfo), action, message.getClass().getSimpleName(),
-                            opaqueId());
+                    logEntry.with(INDICES_FIELD_NAME, toQuotedJsonArray(indices.get()));
                 }
+                opaqueId(threadContext, logEntry);
+                logger.info(logEntry);
             }
         }
     }
 
     @Override
-    public void tamperedRequest(User user, String action, TransportMessage request) {
+    public void tamperedRequest(User user, String action, TransportMessage message) {
         if (events.contains(TAMPERED_REQUEST)) {
-            final Optional<String[]> indices = indices(request);
+            final Optional<String[]> indices = indices(message);
             if (eventFilterPolicyRegistry.ignorePredicate()
                     .test(new AuditEventMetaInfo(Optional.of(user), Optional.empty(), Optional.empty(), indices)) == false) {
-                final LocalNodeInfo localNodeInfo = this.localNodeInfo;
+                final StringMapMessage logEntry = new StringMapMessage(this.entryCommonFields.commonFields);
+                logEntry.with(EVENT_TYPE_FIELD_NAME, TRANSPORT_ORIGIN_FIELD_VALUE)
+                        .with(EVENT_ACTION_FIELD_NAME, "tampered_request")
+                        .with(ACTION_FIELD_NAME, action)
+                        .with(REQUEST_NAME_FIELD_NAME, message.getClass().getSimpleName());
+                restOrTransportOrigin(message, threadContext, logEntry);
+                principal(user, logEntry);
                 if (indices.isPresent()) {
-                    logger.info("{}[transport] [tampered_request]\t{}, {}, action=[{}], indices=[{}], request=[{}]{}",
-                            localNodeInfo.prefix, originAttributes(threadContext, request, localNodeInfo), principal(user), action,
-                            arrayToCommaDelimitedString(indices.get()), request.getClass().getSimpleName(), opaqueId());
-                } else {
-                    logger.info("{}[transport] [tampered_request]\t{}, {}, action=[{}], request=[{}]{}", localNodeInfo.prefix,
-                            originAttributes(threadContext, request, localNodeInfo), principal(user), action,
-                            request.getClass().getSimpleName(), opaqueId());
+                    logEntry.with(INDICES_FIELD_NAME, toQuotedJsonArray(indices.get()));
```

Comment:
There is no way to have the log4j2 appender do the formatting for us?

Reply:
Unfortunately no ☹️ 
[Log4j2 layout pattern functions](https://logging.apache.org/log4j/2.x/manual/layouts.html#PatternLayout) are fairly low level, comparable to printf ones. What we need is "format this as a JSON array field value" (escape then quote every value and concatenate with ,), but what we have instead is just the JSON escape function.
I disliked this too, not being able to have a single point were we handle the formatting, but I dislike more the comma separated strings because we are departing from JSON, the motive for all of this.
Also, array values are the only exception, we are not going to add others constructs that need formatting in the java code.